### ISSUE TITLE
AsyncSeq.cache fails with SynchronizationLockException

### ISSUE THREAD

**Issue by ovidiudeac** [fsprojects/FSharp.Control.AsyncSeq#33]
The exception SynchronizationLockException is thrown when using AsyncSeq.cache

Build info:
- .NET45.
- FSharp.ControlAsyncSeq v2.0.1
- FSharp.Core 4.3.0.0

Please note that the issue reproduces with v2.0.0 but it doesn't reproduce with v1.15.0

Exception details:
```
System.Threading.SynchronizationLockException was unhandled by user code
  HResult=-2146233064
  Message=Object synchronization method was called from an unsynchronized block of code.
  Source=FSharp.Control.AsyncSeq
  StackTrace:
       at FSharp.Control.AsyncSeq.cache@816-6.Invoke(Unit unitVar) in C:\GitHub\dsyme\FSharp.Control.AsyncSeq\src\FSharp.Control.AsyncSeq\AsyncSeq.fs:line 817
       at FSharp.Control.AsyncSeq.tryFinally@326-1.System-IDisposable-Dispose() in C:\GitHub\dsyme\FSharp.Control.AsyncSeq\src\FSharp.Control.AsyncSeq\AsyncSeq.fs:line 351
       at FSharp.Control.AsyncSeq.tryFinally@337-3.Invoke(FSharpOption`1 _arg1) in C:\GitHub\dsyme\FSharp.Control.AsyncSeq\src\FSharp.Control.AsyncSeq\AsyncSeq.fs:line 340
       at Microsoft.FSharp.Control.AsyncBuilderImpl.args@753.Invoke(a a)
  InnerException: 
```
This question was also raised on [stack overflow](http://stackoverflow.com/questions/32094543/synchronizationlockexception-in-asyncseq-cache)

The sample code is listed below. 
```
open FSharp.Control

[<EntryPoint>]
let main argv = 
    let asyncOp x =
        async {
            do! Async.Sleep x
            return x
        }

    let chooser x = if x > 1000 then Some x else None

    let execute (input : AsyncSeq<int>) =
        async {
            printfn "First result... Expect 3s sleep"
            let! r1 = AsyncSeq.tryPick chooser input
            printfn "Computing second result. Expect no sleep"
            let! r2 = AsyncSeq.tryPick chooser input
            printfn "Done"
            return (r1,r2)
        }

    printfn "Starting..."

    let v =
        [1000;2000;3000]
        |> AsyncSeq.ofSeq
        |> AsyncSeq.mapAsync asyncOp
        |> AsyncSeq.cache
        |> execute
        |> Async.RunSynchronously


    printfn "results: = %A" v
    do System.Console.ReadKey() |> ignore
    0
```

**Comment by buybackoff** [comment]
I played with this a little. The `SynchronizationLockException` could be fixed easily by changing the `Monitor` to `ManualResetEventSlim` or `SemaphoreSlim(`1,1`). Now the monitor is exited from a different thread and such exception is documented on the `Monitor` MSDN page.  The build-in test for `cache` passes with the changes and the exception goes away in the example, but the example then hangs in a endless loop. Couldn't fix that last night.

**Comment by eulerfx** [comment]
Thanks, created a failing test for this:

```fsharp
[<Test>]
let ``Async.cache should work``() =  
  let expected = List.init 1000 id
  let effects = ref 0
  let s = asyncSeq {
    for item in expected do      
      yield item
      do! Async.Sleep 1
      incr effects
  }
  let cached = s |> AsyncSeq.cache
  let actual = cached |> AsyncSeq.toList
  let actual = cached |> AsyncSeq.toList
  Assert.True((expected = actual), "cached sequence was different from the original")
  Assert.True((!effects = expected.Length))
```

It fails when the `Async.Sleep` is used after `yield`.

One more thing to consider is the waiting strategy. While `ManualResetEventSlim` works, just like `Monitor` it involves blocking a thread during wait. `ManualResetEventSlim` has a lazy reference to a `WaitHandle` which can be used to wait in a non-blocking way, but referencing that defeats the purpose of the `ManualResetEventSlim` over simply using `ManualResetEvent`. Ideally, we would have an async based `ManualResetEventSlim` which spins before resorting to the `WaitHandle` after which it waits in a non-blocking way with `Async.AwaitWaitHandle`.

Two options at this point are 1) `ManualResetEventSlim` with the blocking `Wait` which spins, this does show clear performance improvements over the non-blocking wait using `WaitHandle` in many scenarios and 2) `ManualResetEvent` with the non-blocking wait, which can make more efficient use of threads, but is less performant in the case when the wait time is short.

**Comment by buybackoff** [comment]
SemaphoreSlim has async wait method, with 1,1 it is effectively MRE

**Comment by buybackoff** [comment]
Async.Await... any non F# things is usually slow. We could get TaskAwaiter
from Semaphore.WaitAsync and move the current Try/Finally logic inside
OnCompletion action. With this, we will also avoid the expensive
Try/Finally of AsyncSeq altogether. Writing from mobile, could not post
code now, but I used this pattern recently and performance is the fastest
one could get from any async machinery in .NET

**Comment by eulerfx** [comment]
I've seen `Semaphore.WaitAsync` have similar overhead to just waiting via `WaitHandle`. In some cases, simply doing a `Semaphore.Wait` is faster.

Not quite sure what you mean by getting rid of the try/finally - do you have an example of what you have in mind?

**Comment by buybackoff** [comment]
Will post later

**Comment by buybackoff** [comment]
The idea was to use this pattern:
```
              let task = semaphore.WaitAsync()
              let awaiter = task.GetAwaiter()
              awaiter.OnCompleted(fun _ ->
                if i >= cache.Count then 
                    let! move = ie.MoveNext()
                    cache.Add(move)
                    iRef := i + 1
                Diagnostics.Debug.Assert(semaphore.CurrentCount = 0)
                semaphore.Release() |> ignore
              )
```
Here try/finally becomes an F# language construct instead of AsynSeq builder method. But it doesn't work now because `let! move = ie.MoveNext()` cannot be called inside OnCompleted. Even if it worked, AsynSeq is built around `async` and we would have to cross the TPL/async boundary later. Recently I worked with an API that is build on TPL Tasks (basically it is TaskSeq, an async sequence on top of Tasks), but the project was in F#. I noticed a huge performance impact of Async.AwaitTask and Async.StartAsTask compared to TPL, and my solution was to remove Async completely ([sample](https://github.com/buybackoff/FSharpAsyncVsTPL/blob/master/FSharpTasks/TaskBuilder.fs#L64)). Here it won't work, and I hope that `Async.AwaitWaitHandle` doesn't add unnecessary overhead if performance matters that much.

**Comment by eulerfx** [comment]
@dsyme I think this issue can be closed re PR #34, however might be worth having a separate discussion wrt waiting as noted: https://github.com/fsprojects/FSharp.Control.AsyncSeq/issues/33#issuecomment-137778655

This is more related to Async than AsyncSeq though.

**Comment by dsyme** [comment]
OK, yes, let's close this and add a new issue re perf if needed
